### PR TITLE
Issue/jperf 445 sudo dind

### PR DESCRIPTION
**Pull request by dagguh**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -29,14 +27,14 @@ configurations.all {
 }
 
 dependencies {
+    api("com.atlassian.performance.tools:ssh:[2.0.0,3.0.0)")
+    api("com.github.docker-java:docker-java:3.1.1")
     implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlinVersion")
-    implementation("org.testcontainers:testcontainers:1.10.5")
```
Comment (dagguh) [inline build.gradle.kts:33]: I can try to shrink the scope if it's possible to implement DIND and sudo via `testcontainers`. I could leave out the autocloseables as well and stick to the existing coarse-grained cleanup mechanisms (leaving leftovers if there's an exception somewhere in the middle of `ssh-ubuntu`.

Comment (dagguh) [reply build.gradle.kts:33]: The upshot is: this is already tested in VUs and has less external dependencies (like Docker publishing).
Locally I rewrote infra for this (basically a find and replace) and it's all green and still fast.

File: .editorconfig [hunk 2]
```diff
@@ -1,2 +1,3 @@
-[*.{kt,kts}]
-indent_size = 4
\ No newline at end of file
+[*.{kt, kts}]
+indent_size = 4
+insert_final_newline = true
```
Comment (sebapawlak) [inline .editorconfig:3]: Noice

**Review by sebapawlak — APPROVED** [review]

File: build.gradle.kts [hunk 3]
```diff
@@ -29,14 +27,14 @@ configurations.all {
 }
 
 dependencies {
+    api("com.atlassian.performance.tools:ssh:[2.0.0,3.0.0)")
```
Comment (wyrzyk) [inline build.gradle.kts:30]: Do we have to add an ssh client to the ssh-ubuntu lib?

File: src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SshUbuntuContainer.kt [hunk 4]
```diff
@@ -1,59 +1,16 @@
...
-class SshUbuntuContainer {
-    internal companion object {
-        private val SSH_PORT = 22
-    }
+/**
+ * @param [ssh] Connects via SSH to an Ubuntu with `sudo`.
+ * @param [ports] Binds container ports to host address space.
+ * @param [peerIp] Addresses the Ubuntu within the Docker network.
+ */
+class SshUbuntuContainer internal constructor(
```
Comment (wyrzyk) [inline src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SshUbuntuContainer.kt:11]: Internal constructor means nothing to the java consumer of the library.

File: src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SshUbuntuImage.kt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package com.atlassian.performance.tools.sshubuntu.api
+
+interface SshUbuntuImage {
```
Comment (wyrzyk) [inline src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SshUbuntuImage.kt:3]: Is it an API or SPI? Can I use own implementations as a user of the library?

File: src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SudoSshUbuntuImage.kt [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    override fun <T> runInUbuntu(
+        lambda: (SshUbuntuContainer) -> T
+    ): T {
+        docker
+            .pullImageCmd("rastasheep/ubuntu-sshd")
+            .withTag("16.04")
+            .exec(PullImageResultCallback())
+            .awaitCompletion()
+        return docker
+            .createContainerCmd("rastasheep/ubuntu-sshd:16.04")
```
Comment (wyrzyk) [inline src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SudoSshUbuntuImage.kt:37]: What license does the image use?

File: src/test/kotlin/com/atlassian/performance/tools/sshubuntu/api/SudoSshUbuntuImageIT.kt [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import org.junit.Test
+import java.util.*
+
+class SudoSshUbuntuImageIT {
+
+    @Test
+    fun shouldStartUbuntu() {
+        val docker = DockerClientBuilder.getInstance().build()
+        val osName = docker
+            .createNetworkCmd()
```
Comment (wyrzyk) [inline src/test/kotlin/com/atlassian/performance/tools/sshubuntu/api/SudoSshUbuntuImageIT.kt:15]: Is the low-level knowledge about docker networks needed to use the library?

File: src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/docker/ConnectedContainer.kt [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    override fun close() {
+        docker
+            .disconnectFromNetworkCmd()
+            .withContainerId(containerId)
+            .withNetworkId(networkId)
+            .exec()
+    }
+}
+
+fun ConnectToNetworkCmd.execAsResource(
```
Comment (wyrzyk) [inline src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/docker/ConnectedContainer.kt:21]: Does it work from java?

File: build.gradle.kts (hunk 3 above)
Comment (dagguh) [reply build.gradle.kts:30]: Pros:
* the consumer can immediately interact with the Ubuntu via SSH, no need to translate APIs
* the consumer doesn't have to search for and decide on an SSH client library
* ❗️ we can preinstall `sudo` as a part of API

Cons:
* bigger dependency tree - more conflicts to resolve (unless we start using ranges)
* doesn't let users choose their own SSH client lib

File: src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/docker/ConnectedContainer.kt (hunk 8 above)
Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/docker/ConnectedContainer.kt:21]: Yes, but it's not as syntactically sweet as in Kotlin:
```java
ConnectedContainerKt.execAsResource(cmd, docker)
```

File: src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SudoSshUbuntuImage.kt (hunk 6 above)
Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SudoSshUbuntuImage.kt:37]: Default one, good catch
https://github.com/rastasheep/ubuntu-sshd/issues/25

File: src/test/kotlin/com/atlassian/performance/tools/sshubuntu/api/SudoSshUbuntuImageIT.kt (hunk 7 above)
Comment (dagguh) [reply src/test/kotlin/com/atlassian/performance/tools/sshubuntu/api/SudoSshUbuntuImageIT.kt:15]: In this form it is. Adding network optionality complicated the implementation.

Pros:
* simpler, non-branching implementation
* exposes the consumer to the real complexity of networks: internal/external addresses/ports - necessary context when dealing with browsers connecting to servers (VU tests) or servers polling themselves locally (infra upgrade REST polling)

Cons:
* it's extra complexity for the users who do not care about connecting multiple containers

Comment (dagguh) [reply src/test/kotlin/com/atlassian/performance/tools/sshubuntu/api/SudoSshUbuntuImageIT.kt:15]: We definitely have to leave the option for networks, but we might invest into making the usage easier for simpler use cases. I'm not sure if it's worth it. I used this code in multiple tests already and the network creation was easily hidden by a local reused method. Perhaps we should wait and see if it's actually a problem.

File: src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SshUbuntuImage.kt (hunk 5 above)
Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SshUbuntuImage.kt:3]: Initially there was no interface at all. I added it to mimic the original `ssh-ubuntu` design.
Perhaps it's unnecessary. TBH I'm not sure about the value here.

File: src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SshUbuntuContainer.kt (hunk 4 above)
Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SshUbuntuContainer.kt:11]: True. Hmm, so we make it public and wait and see if we need to add fields later?

**Comment by dagguh** [comment]
Perhaps this PR is overkill for the purposes of `sudo` and Docker in Docker. I opened this, because I already tested it in other places, didn't want my source to go to waste and didn't want to think too hard about it, because I was in a rush. But that reasoning is suboptimal for the community.

Here's an alternative, which is not a shotgun surgery: https://github.com/atlassian/ssh-ubuntu/pull/6
It changes fewer things and yields a more powerful API.

File: src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SudoSshUbuntuImage.kt (hunk 6 above)
Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SudoSshUbuntuImage.kt:37]: Wait a minute...
https://github.com/atlassian/ssh-ubuntu/blob/release-0.1.0/docker/Dockerfile
Currently we're using the same base

File: build.gradle.kts (hunk 3 above)
Comment (wyrzyk) [reply build.gradle.kts:30]: We could avoid the cons and keep pros by introducing another module.

File: src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SshUbuntuContainer.kt (hunk 4 above)
Comment (wyrzyk) [reply src/main/kotlin/com/atlassian/performance/tools/sshubuntu/api/SshUbuntuContainer.kt:11]: We were experimenting with private constructor + factory method. The function will be visible from java, but the user will be warned.

We can also make it public, I'm not insisting.

**Review by wyrzyk — COMMENTED** [review]
The PR changes API significantly and adds more requirements that we had before. If you'd like to push the change forward, please take care to upgrade it in modules that already depend on it. For `ssh` module the new API will make the lib unusable.

**Comment by dagguh** [comment]
This PR is closed, because https://github.com/atlassian/ssh-ubuntu/pull/6 has won the competition and got merged

**Pull request closed without merging**